MA-3065: port mongoengine to pymongo 3.x

bremac opened the pull request:
The new version should be backwards-compatible with the old version, aside from a few minor changes:

* Read / write concern can no longer be controlled on a per-operation basis. Use `pymongo.collection.Collection.with_options` instead.
* Indexes with drop_dups are no longer supported.
* The `full_response` option has been removed from `QuerySet.modify(...)` since pymongo no longer supports it.

Commit:
MA-3065: port mongoengine to pymongo 3.x

The new version should be backwards-compatible with the old version, aside from
a few minor changes:

* Read / write concern can no longer be controlled on a per-operation basis.
  Use pymongo.collection.Collection.with_options instead.
* Indexes with drop_dups are no longer supported.
* The `full_response` option has been removed from QuerySet.modify(...) since
  pymongo no longer supports it.

Commit:
MA-3065: remove redundant _prune_changed_fields method

_prune_changed_fields was introduced to work around an error in the save path;
it can now be safely removed since the save path has been corrected.

mklein0 commented on mongoengine/document.py:
```diff
@@ -300,11 +288,13 @@ def save(self, force_insert=False, validate=True, clean=True,
+                object_id = collection.insert_one(doc).inserted_id
+            elif created:
+                # If created is true then the delta will not be in sync with the
+                # document; we must upsert the document to preserve all fields.
```
Does this operation take the union of the existing data, or does a full document replacement?

bremac replied:
Good question. It does a replacement, which is the same as the old behavior using `collection.save(...)`:

```python
>>> from pymongo import MongoClient
>>> client = MongoClient()
>>> collection = client['test']['foo']
>>> collection.save({"_id": 1, "foo": 2, "bar": 3})
1
>>> collection.save({"_id": 1, "foo": 4, "baz": 5})
1
>>> list(collection.find())
[{u'_id': 1, u'baz': 5, u'foo': 4}]
```

mklein0 replied:
Should update the comment because it is unclear what is being preserved.

bremac replied:
@mklein0: please let me know if the new version of the comment makes more sense.

bremac committed:
MA-3065: improve comment in `Document.save()`

Clarify what is out of sync and why we need to upsert if an object is marked `created`.
